### DISCUSSION TITLE
Background Script Execution output to sn-scriptsync

### DISCUSSION THREAD

**Discussion by abhishekg999** [arnoudkooi/SN-Utils#480] [Ideas]
Hi, this would require changes in both SN-Utils as well as the VSCode extension, but the idea would be bg script output back to VSCode when using the "Execute Background script" action. Personally, its quite convenient to work in VSCode when testing / working in BGscripts, and it would be pretty nice to be able to work entirely in the editor during this process.

The existing framework in `scriptsync.js` and `extension.ts` seems to have the necessary foundations to implement this. The output can be written to a vscode Output Channel using `vscode.window.createOutputChannel`, and the content can simply be the information in `#result > pre`.

**Comment by abhishekg999** [comment]
I implemented a POC in the following two forks:

https://github.com/abhishekg999/sn-scriptsync
https://github.com/abhishekg999/ServiceNow-Utils

![image](https://github.com/arnoudkooi/ServiceNow-Utils/assets/32551454/2d254b64-9784-4463-8af1-fc6cea6c324b)


The style and everything can obviously be customized to how the community believes is best.
The basic principle is to use a slightly modified version of the `StartBackgroundScript` method instead of clicking the button. Currently as implemented it doesn't use the settings specified in the website, but this can all be synced. It can also be synced to write the data back in the website as well. 

```js
setTimeout(() => {
  document.querySelector('input[name=runscript]').click()
  snuStartGlobalBackgroundScript(editor.getModel().getValue(), e.detail.content.instance, function (response) {
	  console.log("BG script executed");
	  console.log(response);
	snuScriptSyncPostData({action: "writeResponseToTab", instance: e.detail.content.instance, response: response});

  });
  // document.querySelector('input[name=runscript]').click()
}, 500);
```

It then sends this back via the SyncPostData. In the VSCode extension, we simply catch this action, and write it in the Output channel.
```js
const outputChannel = vscode.window.createOutputChannel('Background Execution');
outputChannel.clear();
outputChannel.appendLine(jsn.response);
outputChannel.show();
```

**Comment by abhishekg999** [comment]
Updated to use VSCode Webview Panel. This keeps the code much cleaner as well and removes the he dependency.

![image](https://github.com/arnoudkooi/ServiceNow-Utils/assets/32551454/b74e5e3c-05cc-4d48-acf3-20cedd550ffa)

**Comment by arnoudkooi** [comment]
Thanks for checking in on SN Devs Slack
**sn-scriptsync**
- Please revert the package.json / lock and revert indentation so changes are clearly readable.
- Direct on starting bg execution the webview should be visible with a message that the script is running. preferable with a stopwatch like in platform.
- Creating new BG script should be handled in VS Code, not dependent on The Mirror function
- The eventname to start execution should be `executeBackgroundScript`
- The event that comes back `responseFromBackgroundScript`
- Optionally 

**SN Utils**
There is no need to route the message to sys.script.do and change the logic in bgscript.js . 
The event can be directly handled in scriptsync.js via method `snuStartBackgroundScript`
I will take care of this update with the mentioned event names. 
I will also make below change so in the instance object you can pass a scope where in witch to execute the script.
`sys_scope: instance?.scope || "e24e9692d7702100738dc0da9e6103dd",
`

**Reply by arnoudkooi** [reply]
I just did a commit that implemented above in scriptsync.js

**Reply by abhishekg999** [reply]
Gotcha thank you, will refork here and make respective changes to extension.

**Reply by abhishekg999** [reply]
The latest commit has a small error on scriptsync.js:707. 
Should just be `data = ...` rather than `let data = `. This causes the [Mirror in sn-scriptsync] link to fail and won't prompt the Accept connection box.

![image](https://github.com/arnoudkooi/ServiceNow-Utils/assets/32551454/368425d8-725e-44a7-9450-4151e0dc547f)

**Comment by abhishekg999** [comment]
https://github.com/arnoudkooi/ServiceNow-Utils/assets/32551454/84caebe6-0d49-4654-9b06-17e64f990dd4


Removed most changes in ServiceNow-Utils directory, made the `bgScriptExecute` function send the `action = 'executeBackgroundScript'`, and initialize the VSCode window to start a timer. The timer is refreshed on a 100ms interval same as the site. When it gets a response, it will rewrite the window with the new content.

**Comment by arnoudkooi** [comment]
Good, maybe just make the naming of the functions and variables consistent like below and I'm ok with a scriptsync PR, thanks!
```
webViewPanel
initializeWebViewPanelIfNotExists
writeBGScriptStartToWebViewPanel
writeResponseToWebViewPanel
```

**Reply by abhishekg999** [reply]
sounds great will do!
